### PR TITLE
Fix printing of filenames with surrogate escapes

### LINKED ISSUE #768: UnicodeEncodeError when printing a filename with invalid UTF-8 after conversion
If a filename contains invalid UTF-8 sequences, `dangerzone-cli` fails to print the filename of the successfully converted/failed document due to an `UnicodeEncodeError`.

Steps to reproduce (use bash for [ANSI-C](https://www.gnu.org/software/bash/manual/html_node/ANSI_002dC-Quoting.html) quoting):

Printing filenames of successfully converted documents:
```bash
curl -sSL 'https://raw.githubusercontent.com/freedomofpress/dangerzone/main/tests/test_docs/sample-pdf.pdf' -o $'sample\xf0.pdf'
dangerzone-cli $'sample\xf0.pdf'
```

<details>
<summary>Output</summary>

```
╭──────────────────────────╮
│           ▄██▄           │
│          ██████          │
│         ███▀▀▀██         │
│        ███   ████        │
│       ███   ██████       │
│      ███   ▀▀▀▀████      │
│     ███████  ▄██████     │
│    ███████ ▄█████████    │
│   ████████████████████   │
│    ▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀    │
│                          │
│    Dangerzone v0.6.0     │
│ https://dangerzone.rocks │
╰──────────────────────────╯
Assigning ID 'D-TevQ' to doc '/tmp/sample_.pdf'

Converting document to safe PDF
> /usr/bin/podman run --network none -u dangerzone --log-driver none --security-opt no-new-privileges --userns keep-id --cap-drop all --rm -i dangerzone.rocks/dangerzone /usr/bin/python3 -m dangerzone.conversion.doc_to_pixels
[doc D-TevQ] 0% Converting page 1/4 to pixels
[doc D-TevQ] 12% Converting page 2/4 to pixels
[doc D-TevQ] 24% Converting page 3/4 to pixels
[doc D-TevQ] 36% Converting page 4/4 to pixels
[doc D-TevQ] 49% Converted document to pixels
> /usr/bin/podman run --network none -u dangerzone --log-driver none --security-opt no-new-privileges --userns keep-id --cap-drop all --rm -i -v /tmp/tmphot68xih:/safezone:Z -e OCR=0 -e OCR_LANGUAGE=None dangerzone.rocks/dangerzone /usr/bin/python3 -m dangerzone.conversion.pixels_to_pdf
[doc D-TevQ] 50% Converting page 1/4 from pixels to PDF
[doc D-TevQ] 61% Converting page 2/4 from pixels to PDF
[doc D-TevQ] 72% Converting page 3/4 from pixels to PDF
[doc D-TevQ] 83% Converting page 4/4 from pixels to PDF
[doc D-TevQ] 100% Safe PDF created

Safe PDF(s) created successfully
Traceback (most recent call last):
  File "/usr/bin/dangerzone-cli", line 33, in <module>
    sys.exit(load_entry_point('dangerzone==0.6.0', 'console_scripts', 'dangerzone-cli')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/dangerzone/errors.py", line 103, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/dangerzone/cli.py", line 101, in cli_main
    click.echo(document.output_filename)
  File "/usr/lib/python3.11/site-packages/click/utils.py", line 318, in echo
    file.write(out)  # type: ignore
    ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/colorama/ansitowin32.py", line 47, in write
    self.__convertor.write(text)
  File "/usr/lib/python3.11/site-packages/colorama/ansitowin32.py", line 179, in write
    self.wrapped.write(text)
UnicodeEncodeError: 'utf-8' codec can't encode character '\udcf0' in position 11: surrogates not allowed
```

</details>

Printing filenames of documents that failed to convert:
```bash
curl -sSL 'https://raw.githubusercontent.com/freedomofpress/dangerzone/main/tests/test_docs/sample_bad_pdf.pdf' -o $'sample-bad\xf0.pdf'
dangerzone-cli $'sample-bad\xf0.pdf'
```

<details>
<summary>Output</summary>

```
╭──────────────────────────╮
│           ▄██▄           │
│          ██████          │
│         ███▀▀▀██         │
│        ███   ████        │
│       ███   ██████       │
│      ███   ▀▀▀▀████      │
│     ███████  ▄██████     │
│    ███████ ▄█████████    │
│   ████████████████████   │
│    ▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀    │
│                          │
│    Dangerzone v0.6.0     │
│ https://dangerzone.rocks │
╰──────────────────────────╯
Assigning ID 'UWTbXN' to doc '/tmp/sample-bad_.pdf'

Converting document to safe PDF
> /usr/bin/podman run --network none -u dangerzone --log-driver none --security-opt no-new-privileges --userns keep-id --cap-drop all --rm -i dangerzone.rocks/dangerzone /usr/bin/python3 -m dangerzone.conversion.doc_to_pixels
ERROR [doc UWTbXN] 0% The document format is not supported

Failed to convert document(s)
Traceback (most recent call last):
  File "/usr/bin/dangerzone-cli", line 33, in <module>
    sys.exit(load_entry_point('dangerzone==0.6.0', 'console_scripts', 'dangerzone-cli')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/dangerzone/errors.py", line 103, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/dangerzone/cli.py", line 111, in cli_main
    click.echo(document.input_filename)
  File "/usr/lib/python3.11/site-packages/click/utils.py", line 318, in echo
    file.write(out)  # type: ignore
    ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/colorama/ansitowin32.py", line 47, in write
    self.__convertor.write(text)
  File "/usr/lib/python3.11/site-packages/colorama/ansitowin32.py", line 179, in write
    self.wrapped.write(text)
UnicodeEncodeError: 'utf-8' codec can't encode character '\udcf0' in position 15: surrogates not allowed
```

</details>

Tested on v0.6.0 on Arch Linux. I was unable to reproduce it via GUI as I was not able to select files with invalid UTF-8 sequences in the filename in the file dialog (such files are not displayed - possible Qt issue?). _Update_: GUI is not affected if passing the filename when starting `dangerzone`.

Since the filenames are printed at the end after all conversions, IIUC it does not impact the conversions, but the exception prevents the filename and and any subsequent filenames from being printed, and in case all documents were converted successfully it also impacts the exit code.

### PR DESCRIPTION
**Pull request by naglis**
On Unix systems a filename can be a sequence of bytes that is not valid UTF-8. Python uses[1] surrogate escapes to allow to decode such filenames to Unicode (bytes that cannot be decoded are replaced by a surrogate; upon encoding the surrogate is converted to the original byte).

From `click` docs[2]:

> Invalid bytes or surrogate escapes will raise an error when written to a stream with `errors="strict"`. This will typically happen with `stdout` when the locale is something like `en_GB.UTF-8`.

~To fix that, we use `click.format_filename`[2] before printing the filenames to `stdout` so that surrogate escapes are replaced by �.~ Update: it was decided (see comment https://github.com/freedomofpress/dangerzone/pull/769#discussion_r1557187002) to instead use `replace_control_chars()` and also update its implementation to use Unicode General Category values to decide which characters to replace.

Fixes #768

[1]: https://peps.python.org/pep-0383/
[2]: https://click.palletsprojects.com/en/8.1.x/api/#click.format_filename

### CODE REVIEW

File: dangerzone/cli.py [hunk 1]
```diff
@@ -108,7 +108,7 @@ def cli_main(
     if documents_failed != []:
         print_header("Failed to convert document(s)")
         for document in documents_failed:
-            click.echo(document.input_filename)
+            click.echo(click.format_filename(document.input_filename))
```
[the remark this answers is not in the document]
Comment (naglis) [reply dangerzone/cli.py:111]: > I'd like us to have a sanitation function that does the above

Sorry, please correct if I misunderstood - I should be using `utils.replace_control_chars()` instead of `click.format_filename()`? Or there should be a separate sanitation function added that does what `click.format_filename()` does but is not related to Click?

[the remark this answers is not in the document]
Comment (apyrgio) [reply dangerzone/cli.py:111]: Whoops, sorry, missed your comment. You're right, I was not clear because I don't have a clear solution to this problem yet. Here's what I think:

1. We have this sanitation need both in CLI code, and in library code. So we definitely need a Click-agnostic function, for consistency.
2. We already have `utils.replace_control_chars()`, but it replaces every non-printable character with "_". This is a bad solution for non-english filenames, but we tolerate it for now, for security reasons.
3. `click.format_filename()` deals with invalid bytes, but does not replace control characters. You can check the following in a Python interpreter, and see that your terminal with interpret the ANSI codes:

      ```python
      print(click.format_filename("\033[31;1;4m BaD TeΧt \u200d ❌ \033[0m"))
      ```

So, the ideal solution would be to have a non-Click function that can:
* Replace control characters AND invalid bytes with a placeholder (� is nice).
* Retain valid Unicode letters that do not belong to the English alphabet.

If you have some time, I would suggest looking into [`unicodedata.category`](https://docs.python.org/3/library/unicodedata.html#unicodedata.category), and the associated [General Category values](https://www.unicode.org/reports/tr44/#General_Category_Values). I think that if we update `util.replace_control_chars()` to replace every Unicode character that belongs to the `C*` / `Zl` / `Zp` categories with �, we would have a pretty good solution.

File: tests/__init__.py [hunk 2]
```diff
@@ -113,11 +113,12 @@ def uncommon_text() -> str:
...
     """
-    return "\033[31;1;4m BaD TeΧt \u200d ❌ \033[0m"
+    return "\033[31;1;4m BaD TeΧt \u200d ❌ \udcf0 \033[0m"
 
 
 @pytest.fixture
 def sanitized_text() -> str:
     """Return a sanitized version of the uncommon_text."""
-    return "_[31;1;4m BaD Te_t _ _ _[0m"
+    return "�[31;1;4m BaD TeΧt � ❌ � �[0m"
```
Comment (naglis) [inline tests/__init__.py:124]: When using General Categories, emojis (or at leas the ❌ emoji) are categorized as So (Other symbol) and thus are not replaced, but I've kept in the test fixtures. Please let me know if it is ok to remove the emoji from the fixtures.

File: tests/test_cli.py [hunk 3]
```diff
@@ -218,6 +225,17 @@ def test_output_filename_spaces(self, sample_pdf: str) -> None:
         result = self.run_cli([sample_pdf, "--output-filename", output_filename])
         result.assert_success()
 
+    @pytest.mark.skipif(
+        platform.system() == "Darwin", reason="filename not supported on macOS"
```
Comment (naglis) [inline tests/test_cli.py:229]: macOS does [not seem to support](https://github.com/freedomofpress/dangerzone/actions/runs/8645853053/job/23703882932#step:6:446) invalid UTF-8 sequences in filenames, so macOs is skipped.

File: dangerzone/cli.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (naglis) [reply dangerzone/cli.py]: I've updated `replace_control_chars` to use General Category values in 3212846c82de60083b4c75a810ac1f28f2238110.

[the remark this answers is not in the document]
Comment (apyrgio) [reply dangerzone/cli.py]: Yeap, I think it's awesome, thanks for the change!

File: tests/__init__.py (hunk 2 above)
Comment (apyrgio) [reply tests/__init__.py:124]: I think it's ok. We can extend the docstring of the `sanitized_text` fixture, to explain why some Unicode characters are sanitized, and others are not.

File: tests/test_cli.py (hunk 3 above)
Comment (apyrgio) [reply tests/test_cli.py:229]: Nice catch, but I'd still like to test as many corner cases in macOS as possible. For this reason, I propose using the `uncommon_text` fixture here, and simply remove the invalid Unicode character. The rest of the tests should ideally work.

Also, this would be a good opportunity to merge it with a test we have for filenames with spaces in it (`test_output_filename_spaces`).

**Comment by apyrgio** [comment]
Hey @naglis, I went ahead and refactored your branch a bit, with the following changes:

1. I added a new `uncommon_filename` fixture, which is basically `uncommon_text`, with the exception of the invalid Unicode character that doesn't work in macOS filenames. I use this fixture instead of your string with Unicode surrogate escapes, so that we can have cross-platform tests for files with uncommon Unicode characters in them.
2. I squashed some commits, so that the Git history does not contain `click.format_filename()`.
3. I retained your Git authorship where applicable.
 
_(Also, sorry for force-pushing in your branch. For reference, the original branch is tracked in`fix-printing-filename-naglis` in this repo)_

If you are ok with the changes, I can resolve the conversations and merge this PR in time for 0.6.1. Else, we can discuss more and merge it afterwards. No pressure in any case :slightly_smiling_face:

**Comment by naglis** [comment]
> If you are ok with the changes, I can resolve the conversations and merge this PR in time for 0.6.1. Else, we can discuss more and merge it afterwards. No pressure in any case 🙂

Fine with me, the changes look good. Thanks for collaborating!

**Review by apyrgio — APPROVED** [review]
